lib/SILOptimizer/IPO/GlobalOpt.cpp

```diff
@@ -817,14 +821,19 @@ static LoadInst *getValidLoad(SILInstruction *I, SILValue V) {
       return LI;
   }
 
-  if (auto *SEAI = dyn_cast<StructElementAddrInst>(I)) {
-    if (SEAI->getOperand() == V && SEAI->hasOneUse())
-      return getValidLoad(SEAI->use_begin()->getUser(), SEAI);
+  if (isa<StructElementAddrInst>(I) || isa<TupleElementAddrInst>(I) ||
+      isa<BeginAccessInst>(I)) {
+    auto singleValue = cast<SingleValueInstruction>(I);
+    if (singleValue->getOperand(0) == V && singleValue->hasOneUse())
+      return getValidLoad(singleValue->use_begin()->getUser(), singleValue);
   }
 
-  if (auto *TEAI = dyn_cast<TupleElementAddrInst>(I)) {
-    if (TEAI->getOperand() == V && TEAI->hasOneUse())
-      return getValidLoad(TEAI->use_begin()->getUser(), TEAI);
+  if (auto beginAccess = dyn_cast<BeginAccessInst>(I)) {
+    if (beginAccess->getOperand() == V &&
+        std::distance(beginAccess->use_begin(), beginAccess->use_end()) == 2) {
```

Comment:
begin_access can have multiple end_access, and that should not affect optimization (same is true for other scoped instructions). There should be some utility to get only the non-incidental uses.

Reply:
You're right. I can add a utility for doing this. Should I add it to `BeginAccessInst` or make it a static function?

Reply:
It might be reasonable to expect a single end_access in this particular scenario, but it is definitely a bug to make a transformation sensitive to the use list order.

Reply:
It should be something that any scoped instruction can use. Like begin_apply, begin_borrow, or whatever. I'm not sure if it should be on `SILInstruction` or in `InstructionUtils.h`, but since it uses other helpers in `InstructionUtils` it probable belongs there. You need to decide whether to look at `isEndOfScopeMarker` or `isIncidentalUse`. I think it makes the most sense to look at `isIncidentalUse` here. The definition of an incidental use is one that shouldn't affect optimization of the operation.

Reply:
Thanks for the info/direction. I've added `getSingleNonDebugUser` and was able to combine the `begin_access` specific code with the block above.